src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1128,16 +1034,10 @@ bool ProcessLateMarkedAttributes ()
 			return markOccurred;
 		}
 
-		internal protected void MarkField (FieldReference reference, DependencyInfo reason)
+		internal protected void MarkField (FieldReference reference, MarkingInfo reason)
 		{
 			if (reference.DeclaringType is GenericInstanceType) {
-				Debug.Assert (reason.Kind == DependencyKind.FieldAccess || reason.Kind == DependencyKind.Ldtoken);
-				// Blame the field reference (without actually marking) on the original reason.
-				Tracer.AddDirectDependency (reference, reason, marked: false);
-				MarkType (reference.DeclaringType, new DependencyInfo (DependencyKind.DeclaringType, reference));
-
-				// Blame the field definition that we will resolve on the field reference.
-				reason = new DependencyInfo (DependencyKind.FieldOnGenericInstance, reference);
+				MarkType (reference.DeclaringType, new MarkingInfo (MarkingReason.GenericInstance, reference));
```

Comment:
```csharp
public class G<T> {
  public T field;
}

public class Foo {}

public class C {
  public void M() {
    var g = new G<Foo> ();
    _ = g.field;
  }
}
```
Before the changes here and in GetOriginalType, this creates a dependency graph like:

- `M`
  - FieldAccess -> `G<Foo>.field`
    - DeclaringType (here) -> `G<Foo>`
      - ElementType (in GetOriginalType) -> `G<T>`
      - GenericArgumentType -> `Foo`
    - FieldOnGenericInstance -> `G<T>.field`
      - DeclaringType (in MarkField) -> `G<T>`

With the new logic, it looks like this:

- `M`
  - FieldAccessInstr -> `G<T>.field`
    - DeclaringType (in MarkField) -> `G<T>`
- `G<Foo>.field`
  - GenericInstance -> `G<T>`
- `G<Foo>`
  - GenericArgument -> `Foo`

which includes `G<Foo>.field` or `G<Foo>` in the dependency graph without saying why, and also seems to record `G<T>` as a generic instance of `G<Foo>.field` which doesn't make sense to me. What is the reason for the change?